src/main/java/picard/illumina/ClusterDataToSamConverter.java

```diff
@@ -207,49 +241,86 @@ private SAMRecord createSamRecord(final ReadData readData, final String readName
+            for (int molecularBarcodeIndice : molecularBarcodeIndices) {
+                molecularIndexes.add(new String(cluster.getRead(molecularBarcodeIndice).getBases()).replace('.', 'N'));
+                molecularIndexQualities.add(SAMUtils.phredToFastq(cluster.getRead(molecularBarcodeIndice).getQualities()));
             }
-        } else {
-            molecularIndexes        = Collections.emptyList();
-            molecularIndexQualities = Collections.emptyList();
         }
 
         final SAMRecord firstOfPair = createSamRecord(
-            cluster.getRead(templateIndices[0]), readName, cluster.isPf(), true, unmatchedBarcode, molecularIndexes, molecularIndexQualities);
+                cluster.getRead(templateIndices[0]),
+                readName,
+                cluster.isPf(),
+                true,
+                unmatchedBarcode,
+                barcodeQuality,
+                molecularIndexes,
+                molecularIndexQualities);
         ret.records[0] = firstOfPair;
 
+
         SAMRecord secondOfPair = null;
 
-        if(isPairedEnd) {
-            secondOfPair  = createSamRecord(
-                cluster.getRead(templateIndices[1]), readName, cluster.isPf(), false, unmatchedBarcode, molecularIndexes, molecularIndexQualities);
+        if (isPairedEnd) {
+            secondOfPair = createSamRecord(
+                    cluster.getRead(templateIndices[1]),
+                    readName,
+                    cluster.isPf(),
+                    false,
+                    unmatchedBarcode,
+                    barcodeQuality,
+                    molecularIndexes,
+                    molecularIndexQualities);
             ret.records[1] = secondOfPair;
         }
 
         if (adapterMarker != null) {
             // Clip the read
             if (isPairedEnd) {
                 adapterMarker.adapterTrimIlluminaPairedReads(firstOfPair, secondOfPair);
-            }
-            else {
+            } else {
                 adapterMarker.adapterTrimIlluminaSingleRead(firstOfPair);
             }
         }
         return ret;
     }
+
+    private String getBarcodeQuality(ClusterData cluster) {
+        final StringJoiner barcodeQ = new StringJoiner("~");
+
+        for (int sampleBarcodeIndex : sampleBarcodeIndices) {
+            barcodeQ.add(SAMUtils.phredToFastq(cluster.getRead(sampleBarcodeIndex).getQualities()));
+        }
+        return barcodeQ.toString();
+    }
+
+    private String getUnmatchedBarcode(ClusterData cluster, boolean withSeparator) {
+        final byte[][] barcode = new byte[sampleBarcodeIndices.length][];
+        for (int i = 0; i < sampleBarcodeIndices.length; i++) {
+            barcode[i] = cluster.getRead(sampleBarcodeIndices[i]).getBases();
+        }
+        if (withSeparator) {
+            //TODO: This has a separator, where as in other places we do not use a separator
+            return IlluminaUtil.barcodeSeqsToString(barcode).replace('.', 'N');
```

Comment:
The expression `.replace('.', 'N')` is repeated a few times in this file. It would be better to put it in a method with a meaningful name.

Reply:
yup.